### PR TITLE
Fix to prevent the onNext event going to stale subscription when restart happens in poller

### PR DESCRIPTION
**Pull request by ashwing**
Fix to prevent the onNext event going to stale subscription when restart happens in poller

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/PrefetchRecordsPublisher.java [hunk 1]
```diff
@@ -216,12 +216,15 @@ public void restartFrom(RecordsRetrieved recordsRetrieved) {
         PrefetchRecordsRetrieved prefetchRecordsRetrieved = (PrefetchRecordsRetrieved) recordsRetrieved;
         resetLock.writeLock().lock();
         try {
+            // Reset the demand from ShardConsumer, to prevent this publisher from delivering events to stale RX-Java
+            // Subscriber. Publishing will be unblocked when the demand is communicated by the new Rx-Java subscriber.
+            requestedResponses.set(0);
```
Comment (yasemin-amzn) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/PrefetchRecordsPublisher.java]: this class keep track of so much state, and they all need to be reset at restartFrom call. I'm wondering if pulling out the state tracking variable in one common place, and call reset on that one to start with clean state helps longer term maintainability of this class. What do you think?

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/PrefetchRecordsPublisher.java [hunk 2]
```diff
@@ -296,45 +309,27 @@ private void addArrivedRecordsInput(PrefetchRecordsRetrieved recordsRetrieved) t
...
-            // thread.
-            if (shouldDrainEventOnlyOnAck.get()) {
-                log.debug("{} : Publisher thread takes over the draining control. Queue Size : {}, Demand : {}",
-                        shardId, getRecordsResultQueue.size(), requestedResponses.get());
-                shouldDrainEventOnlyOnAck.set(false);
-            }
         }
     }
 
+    private static boolean canDispatchRecord(PrefetchRecordsRetrieved recordsToDeliver) {
```
Comment (yasemin-amzn) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/PrefetchRecordsPublisher.java]: worth adding a comment here, or in the PrefetchesRecordsRetrieved class about multi-threading, and this flag protecting against delivering records twice.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/PrefetchRecordsPublisher.java [hunk 3]
```diff
@@ -404,7 +404,7 @@ public void run() {
 
         private void makeRetrievalAttempt() {
```
Comment (yasemin-amzn) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/PrefetchRecordsPublisher.java]: something to consider: does this method fit better into session now -- there will be direct access to counters, and waiting etc.
Only when retrievalAttempt call to session returns success, then PrefetchRecordsPublisher would call drainQueue.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/PrefetchRecordsPublisher.java (hunk 2 above)
Comment (ashwing) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/PrefetchRecordsPublisher.java]: done

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/PrefetchRecordsPublisher.java [hunk 4]
```diff
@@ -296,45 +309,27 @@ private void addArrivedRecordsInput(PrefetchRecordsRetrieved recordsRetrieved) t
...
-        }
+        publisherSession.prefetchCounters().added(recordsRetrieved.processRecordsInput);
     }
 
     /**
      * Method to drain the queue based on the demand and the events availability in the queue.
      */
     private synchronized void drainQueueForRequests() {
-        final RecordsRetrieved recordsToDeliver = peekNextResult();
+        final PrefetchRecordsRetrieved recordsToDeliver = (PrefetchRecordsRetrieved) peekNextResult();
```
[the remark this answers is not in the document]
Comment (ashwing) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/PrefetchRecordsPublisher.java]: done

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/PrefetchRecordsPublisher.java [hunk 5]
```diff
@@ -283,7 +296,7 @@ public synchronized void notify(RecordsDeliveryAck recordsDeliveryAck) {
     // Note : Do not make this method synchronous as notify() will not be able to evict any entry from the queue.
     private void addArrivedRecordsInput(PrefetchRecordsRetrieved recordsRetrieved) throws InterruptedException {
         wasReset = false;
-        while (!getRecordsResultQueue.offer(recordsRetrieved, idleMillisBetweenCalls, TimeUnit.MILLISECONDS)) {
+        while (!publisherSession.prefetchRecordsQueue().offer(recordsRetrieved, idleMillisBetweenCalls, TimeUnit.MILLISECONDS)) {
```
[the remark this answers is not in the document]
Comment (ashwing) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/PrefetchRecordsPublisher.java]: done

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/PrefetchRecordsPublisher.java [hunk 6]
```diff
@@ -257,16 +277,9 @@ public synchronized void notify(RecordsDeliveryAck recordsDeliveryAck) {
         // Verify if the ack matches the head of the queue and evict it.
         if (recordsToCheck != null && recordsToCheck.batchUniqueIdentifier()
```
[the remark this answers is not in the document]
Comment (ashwing) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/PrefetchRecordsPublisher.java]: done

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/PrefetchRecordsPublisher.java [hunk 7]
```diff
@@ -181,23 +223,13 @@ private void throwOnIllegalState() {
 
     private RecordsRetrieved peekNextResult() {
         throwOnIllegalState();
-        final PrefetchRecordsRetrieved result = getRecordsResultQueue.peek();
-        return result == null ? result : result.prepareForPublish();
+        return publisherSession.prefetchRecordsQueue().peek();
```
[the remark this answers is not in the document]
Comment (ashwing) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/PrefetchRecordsPublisher.java]: done

**Pull request merged**